### PR TITLE
Initial refactor of the protocol generation logic

### PR DESCRIPTION
**Pull request by patrickherring-TRI**
This PR is an attempt to make protocol generation more intuitive and easier to modify.
- Any template with the right number of steps will be parameterized similarly - new templates with minor modifications are fine
- Additional decision logic can be added for more specific cases
- protocol files can be overwritten now

### CODE REVIEW

File: beep/generate_protocol.py [hunk 1]
```diff
@@ -89,96 +108,88 @@ def generate_protocol_files_from_csv(csv_filename, output_directory=None):
...
-            )
-            diagnostic_params = diag_params_df[
-                diag_params_df["diagnostic_parameter_set"]
-                == protocol_params["diagnostic_parameter_set"]
-            ].squeeze()
+        if ".000" in template:
             template_fullpath = os.path.join(PROCEDURE_TEMPLATE_DIR, template)
-            if protocol_params["project_name"] == "RapidC":
+            template_length = template_detection(template_fullpath)
+            if "diagnostic_parameter_set" in protocol_params:
```
Comment (moorepatrick) [inline beep/generate_protocol.py:125]: Can we put in some comments next to each of these `if`'s and `elif`'s to describe what we are matching.

File: beep/generate_protocol.py [hunk 2]
```diff
@@ -89,96 +108,88 @@ def generate_protocol_files_from_csv(csv_filename, output_directory=None):
...
+                file_generation_failures.append(failure)
+                warnings.warn("Unsupported file template {}, skipping.".format(template))
+                result = "error"
+                continue
+
             filename = "{}.000".format(filename_prefix)
             filename = os.path.join(output_directory, "procedures", filename)
-        elif template == "formationV1.mps":
+
+        elif ".mps" in template and template == "formationV1.mps":
```
Comment (moorepatrick) [inline beep/generate_protocol.py:179]: Some more comments next to these would be helpful too.

**Review by moorepatrick — CHANGES_REQUESTED** [review]
LGTM but a few more comments would be helpful.

File: beep/generate_protocol.py (hunk 1 above)
Comment (patrickherring-TRI) [reply beep/generate_protocol.py:125]: Good idea! Added those now

File: beep/generate_protocol.py (hunk 2 above)
Comment (patrickherring-TRI) [reply beep/generate_protocol.py]: Yup, adding comments to each of the cases

**Review by moorepatrick — APPROVED** [review]
LGTM

**Review by chirranjeevigopal-TRI — APPROVED** [review]
LGTM!

**Pull request merged**